schematic/schemas/json_schema_validation_rule_functions.py

```diff
@@ -0,0 +1,367 @@
+"""
+This module contains functions for interacting with validation rules for JSON Schema creation
+
+JSON Schema docs:
+
+Array: https://json-schema.org/understanding-json-schema/reference/array
+Types: https://json-schema.org/understanding-json-schema/reference/type#type-specific-keywords
+Format: https://json-schema.org/understanding-json-schema/reference/type#format
+Pattern: https://json-schema.org/understanding-json-schema/reference/string#regexp
+Min/max: https://json-schema.org/understanding-json-schema/reference/numeric#range
+
+"""
+
+from dataclasses import dataclass
+from typing import Optional
+from schematic.schemas.constants import JSONSchemaType, ValidationRuleName, RegexModule
+
+
+@dataclass
+class ValidationRule:
+    """
+    This class represents a Schematic validation rule to be used for creating JSON Schemas
+
+    Attributes:
+        name: The name of the validation rule
+        js_type: The JSON Schema type this rule indicates.
+          For example type rules map over to their equivalent JSON Schema type: str -> string
+          Other rules have am implicit type. For example the regex rule maps to the JSON
+            Schema pattern keyword. The pattern keyword requires the type to be string
+        incompatible_rules: Other validation rules this rule can not be paired with
+        parameters: Parameters for the validation rule that need to be collected for the JSON Schema
+    """
+
+    name: ValidationRuleName
+    js_type: Optional[JSONSchemaType]
+    incompatible_rules: list[ValidationRuleName]
+    parameters: Optional[list[str]] = None
+
+
+# A dictionary of current Schematic validation rules
+#   where the keys are name fo the rule in Schematic
+#   and the values are ValidationRule objects
+_VALIDATION_RULES = {
```

Comment:
I am trying to think if there's a better way rather than manually maintaining a list of validation rules, especially when: 
maintaining manual, duplicated, and asymmetric incompatible_rules lists is brittle and error-prone. Especially when:
* Rule A lists Rule B as incompatible, but B doesn’t reciprocate.
* Someone adds a new rule but forgets to update other rules’ incompatible_rules.

Can we try to maintain a map of incompatible rule pair? is that better?

Reply:
I agree, but not sure what the solution is. That said, like I mentioned in a different comment this should be the last time we add any validation rules to be converted into JSON Schema keywords. There just aren't anymore Schematic validation rules that have a direct translation into JSON Schema like `date` or `url`.